File: MicroTDCore/Sources/MicroTDCore/Systems/ProgressionSystem.swift
// ProgressionSystem.swift
// Pure logic system for applying run results to progression profile

import Foundation

public struct ProgressionSystem: Sendable {
    public let rules: ProgressionRules
    
    public init(rules: ProgressionRules = ProgressionRules()) {
        self.rules = rules
    }
    
    /// Apply run summary to profile, returning events and updated profile
    /// This is a PURE function - it does not mutate input in place, but returns usage result
    /// (Or mutating if preferred, but explicit is nice)
    public func applyRun(_ run: RunSummary, to profile: inout ProgressionProfile) -> [ProgressionEvent] {
        var events: [ProgressionEvent] = []
        
        // 1. Calculate and Add XP
        let gainedXP = rules.xpForRun(run)
        if gainedXP > 0 {
            profile.xp += gainedXP
            events.append(.xpGained(amount: gainedXP))
        }
        
        // 2. Check for Level Up
        // Continue leveling up as long as we have enough XP
        while true {
            let threshold = rules.totalXpRequiredToReachNextLevel(from: profile.level)
            if profile.xp >= threshold {
                let oldLevel = profile.level
                profile.level += 1
                // XP is cumulative in this model (Standard RPG style), or resets?
                // "xpRequiredForLevel(1)" = 100.
                // If I have 150, I reach level 2. Do I keep 150? Or 50?
                // "100 * level * level" implies cumulative threshold.
                // L1->L2 needs 100 total. L2->L3 needs 400 total.
                
                events.append(.leveledUp(from: oldLevel, to: profile.level))
                
                // 3. Check for Unlocks
                let newUnlocks = rules.unlocksForLevel(profile.level)
                for unlockID in newUnlocks {
                    if !profile.unlocks.contains(unlockID) {
                        profile.unlocks.insert(unlockID)
                        events.append(.unlocked(id: unlockID))
                    }
                }
                
            } else {
                break
            }
        }
        
        return events
    }
}

File: expo/src/core/systems/ProgressionSystem.ts
// ProgressionSystem.ts
// Pure logic for applying run results to player profile
// Ported from MicroTDCore/Systems/ProgressionSystem.swift

import type { ProgressionProfile, RunSummary, ProgressionEvent } from '../definitions/ProgressionTypes';
import { ProgressionRules } from '../definitions/ProgressionTypes';

/**
 * Pure logic system that processes run summaries and updates player profiles.
 * No I/O — just computation. Persistence is handled by ProfileStore/RunManager.
 */
export class ProgressionSystem {
    private readonly rules: ProgressionRules;

    constructor(rules: ProgressionRules = new ProgressionRules()) {
        this.rules = rules;
    }

    /**
     * Apply a completed run to a profile. Mutates the profile in place and
     * returns events describing what changed (XP gained, level ups, unlocks).
     */
    applyRun(summary: RunSummary, profile: ProgressionProfile): ProgressionEvent[] {
        const events: ProgressionEvent[] = [];

        // Calculate XP earned
        const xpGained = this.rules.calculateXP(summary);
        if (xpGained <= 0) return events;

        // Apply XP
        const previousLevel = profile.level;
        profile.xp += xpGained;

        events.push({
            type: 'xpGained',
            amount: xpGained,
            newTotal: profile.xp,
        });

        // Check for level ups
        const newLevel = this.rules.levelForXP(profile.xp);
        if (newLevel > previousLevel) {
            profile.level = newLevel;
            events.push({
                type: 'leveledUp',
                newLevel: newLevel,
            });

            // Award any unlocks for gained levels
            for (let lvl = previousLevel + 1; lvl <= newLevel; lvl++) {
                const levelUnlocks = this.rules.unlockSchedule
                    .filter(u => u.level === lvl)
                    .map(u => u.itemID);

                for (const itemID of levelUnlocks) {
                    if (!profile.unlocks.has(itemID)) {
                        profile.unlocks.add(itemID);
                        events.push({
                            type: 'unlocked',
                            itemID,
                        });
                    }
                }
            }
        }

        return events;
    }

    /**
     * Reconcile unlocks — ensure profile has all unlocks it should for its level.
     * Used after loading a profile to handle cases where unlocks were missed.
     */
    reconcileUnlocks(profile: ProgressionProfile): void {
        const expectedUnlocks = this.rules.unlocksForLevel(profile.level);
        for (const itemID of expectedUnlocks) {
            profile.unlocks.add(itemID);
        }
    }
}
